### PR TITLE
Add more support in imap nio client

### PR DESCRIPTION
**Pull request by jui8wang**
1. Support Compress command, add handlers to deflate/inflate
2. Support Uid expunge command
3. Support parsing
 - response of Status command to Status object
 - response of ID command to IdResult
4. Use UIDSet or String to support 0 to 2^32 -1 values for UID values
5. Use Config class instead of Property class for ImapAsyncClient.

## Description

1. Support Compress command, add handlers to deflate/inflate
2. Support Uid expunge command
3. Support parsing
 - response of Status command to Status object
 - response of ID command to IdResult
4. Use UIDSet or String to support 0 to 2^32 -1 values for UID values
5. Use Config class instead of Property class for ImapAsyncClient.
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Major release (change is NOT backward compatible with prior release)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java [hunk 1]
```diff
@@ -317,6 +329,20 @@ public void handleIdleEvent(@Nonnull final IdleStateEvent idleEvent) {
             return;
 
         } else if (serverResponse.isTagged()) {
+            if (currentCmd instanceof CompressCommand && serverResponse.isOK()) {
```
Comment (nituyahoo) [inline core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java]: we should avoid looking for command type

File: core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java [hunk 2]
```diff
@@ -317,6 +329,20 @@ public void handleIdleEvent(@Nonnull final IdleStateEvent idleEvent) {
...
+                final ChannelPipeline pipeline = ch.pipeline();
+                final JdkZlibDecoder decoder = new JdkZlibDecoder(ZlibWrapper.NONE);
+                final JdkZlibEncoder encoder = new JdkZlibEncoder(ZlibWrapper.NONE, 5);
+                if (pipeline.get(ImapAsyncClient.SSL_HANDLER) == null) {
+                    // no SSL handler, deflater/enflater has to be first
+                    pipeline.addFirst(ZLIB_DECODER, decoder);
+                    pipeline.addFirst(ZLIB_ENCODER, encoder);
+                } else {
+                    pipeline.addAfter(ImapAsyncClient.SSL_HANDLER, ZLIB_DECODER, decoder);
+                    pipeline.addAfter(ImapAsyncClient.SSL_HANDLER, ZLIB_ENCODER, encoder);
```
Comment (nituyahoo) [inline core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java]: once added do we remove it?

File: core/src/main/java/com/lafaspot/imapnio/async/request/UidFetchCommand.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        this(UIDSet.toString(uidsets), what);
+    }
+
+    /**
+     * Initializes a @{code UidFetchCommand} with the message sequence syntax.
+     *
+     * @param msgIds the message sequence or UID string. For ex:2:4
+     * @param what the data items or macro
+     */
+    public UidFetchCommand(@Nonnull final String msgIds, @Nonnull final String what) {
```
Comment (nituyahoo) [inline core/src/main/java/com/lafaspot/imapnio/async/request/UidFetchCommand.java]: should this be uid and not msgids

File: core/src/main/java/com/lafaspot/imapnio/async/request/UidSearchCommand.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        this(UIDSet.toString(uidsets), term);
+    }
+
+    /**
+     * Initializes a @{code UidSearchCommand} with the msg UID string directly.
+     *
+     * @param msgIds the messages UID string
+     * @param term the search expression tree
+     */
+    private UidSearchCommand(@Nonnull final String msgIds, @Nonnull final SearchTerm term) {
```
Comment (nituyahoo) [inline core/src/main/java/com/lafaspot/imapnio/async/request/UidSearchCommand.java]: uids?

File: core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java (hunk 1 above)
Comment (jui8wang) [reply core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java]: Done by having a method in ImapRequest interface isCompressionRequested().

File: core/src/main/java/com/lafaspot/imapnio/async/client/ImapAsyncClient.java [hunk 5]
```diff
@@ -163,27 +162,26 @@ public ImapAsyncClient(final int numOfThreads) throws SSLException {
...
      * @return the ChannelFuture object
      */
-    public ImapFuture<ImapAsyncSession> createSession(@Nonnull final URI serverUri, @Nonnull final Properties properties,
+    public ImapFuture<ImapAsyncSession> createSession(@Nonnull final URI serverUri, @Nonnull final ImapAsyncSessionConfig config,
             @Nullable final InetSocketAddress localAddress, @Nullable final List<String> sniNames) {
         // ------------------------------------------------------------
         // check if required properties are present
-        properties.putIfAbsent(CONFIG_CONNECTION_TIMEOUT_KEY_MILLIS, DEFAULT_CONNECTION_TIMEOUT_MILLIS);
-        properties.putIfAbsent(CONFIG_IMAP_TIMEOUT_KEY_MILLIS, DEFAULT_IMAP_TIMEOUT_MILLIS);
+        final int connectionTimeMillis = (config.getConnectionTimeoutMillis() <= 0) ? DEFAULT_CONNECTION_TIMEOUT_MILLIS
```
[the remark this answers is not in the document]
Comment (jui8wang) [reply core/src/main/java/com/lafaspot/imapnio/async/client/ImapAsyncClient.java]: Done.

File: core/src/main/java/com/lafaspot/imapnio/async/data/IdResult.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        return params.containsKey(keyName);
+    }
+
+    /**
+     * Returns the value based on the given key name.
+     *
+     * @param keyName the capability name to find
+     * @return the value
+     */
+    public String getValue(@Nonnull final String keyName) {
```
[the remark this answers is not in the document]
Comment (jui8wang) [reply core/src/main/java/com/lafaspot/imapnio/async/data/IdResult.java]: Done and added UT.

File: core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java (hunk 2 above)
Comment (jui8wang) [reply core/src/main/java/com/lafaspot/imapnio/async/internal/ImapAsyncSessionImpl.java]: Discussed.

File: core/src/main/java/com/lafaspot/imapnio/async/request/UidFetchCommand.java (hunk 3 above)
Comment (jui8wang) [reply core/src/main/java/com/lafaspot/imapnio/async/request/UidFetchCommand.java]: Done.

File: core/src/main/java/com/lafaspot/imapnio/async/request/UidSearchCommand.java (hunk 4 above)
Comment (jui8wang) [reply core/src/main/java/com/lafaspot/imapnio/async/request/UidSearchCommand.java]: Done.

**Comment by jui8wang** [comment]
Will submit a new pull request to address the comments.

**Pull request closed without merging**